pkl-core/src/main/java/org/pkl/core/packages/DependencyMetadata.java

```diff
@@ -423,8 +584,129 @@ private void write() throws IOException {
     }
+
+    private void writeAnnotations() throws IOException {
+      jsonWriter.beginArray();
+      for (var annotation : dependencyMetadata.annotations) {
+        writePObject(annotation);
+      }
+      jsonWriter.endArray();
+    }
+
+    private void writePClassInfo(PClassInfo<?> pClassInfo) throws IOException {
+      jsonWriter.beginObject();
+      jsonWriter.name("moduleName").value(pClassInfo.getModuleName());
+      jsonWriter.name("class").value(pClassInfo.getSimpleName());
+      jsonWriter.name("moduleUri").value(pClassInfo.getModuleUri().toString());
+      jsonWriter.endObject();
+    }
+
+    private void writePObject(PObject object) throws IOException {
+      jsonWriter.beginObject();
+      jsonWriter.name("type").value("PObject");
+      jsonWriter.name("classInfo");
+      writePClassInfo(object.getClassInfo());
+      jsonWriter.name("properties");
+      jsonWriter.beginObject();
+      for (var kv : object.getProperties().entrySet()) {
+        jsonWriter.name(kv.getKey());
+        writeGenericObject(kv.getValue());
+      }
+      jsonWriter.endObject();
+      jsonWriter.endObject();
+    }
+
+    private void writeGenericObject(Object value) throws IOException {
+      if (value instanceof PNull) {
+        jsonWriter.nullValue();
+      } else if (value instanceof PObject pObject) {
+        writePObject(pObject);
+      } else if (value instanceof String string) {
+        jsonWriter.value(string);
+      } else if (value instanceof Boolean bool) {
+        jsonWriter.value(bool);
+      } else if (value instanceof Integer num) {
+        jsonWriter.value(num);
+      } else if (value instanceof Long num) {
+        jsonWriter.value(num);
+      } else if (value instanceof Float num) {
+        jsonWriter.value(num);
+      } else if (value instanceof Double num) {
+        jsonWriter.value(num);
+      } else if (value instanceof List<?> list) {
+        jsonWriter.beginArray();
+        for (var v : list) {
+          writeGenericObject(v);
+        }
+        jsonWriter.endArray();
+      } else if (value instanceof Set<?> set) {
+        jsonWriter.beginObject();
+        jsonWriter.name("type").value("Set");
+        jsonWriter.name("value");
+        jsonWriter.beginArray();
+        for (var v : set) {
+          writeGenericObject(v);
+        }
+        jsonWriter.endArray();
+        jsonWriter.endObject();
+      } else if (value instanceof Map<?, ?> map) {
+        jsonWriter.beginObject();
+        jsonWriter.name("type").value("Map");
+        jsonWriter.name("value");
+        jsonWriter.beginObject();
+        for (var kv : map.entrySet()) {
+          var key = kv.getKey();
+          if (key instanceof String s) {
+            jsonWriter.name(s);
+          } else {
+            throw new PklException(
+                "Error serializing annotation for PklProject:\n"
+                    + "  cannot render map with non-string key: "
+                    + key);
+          }
```

Comment:
How about we serialize maps like so?

```
[
  {
    "key": ...
    "value": ...
  }
]
```

Reply:
Good shout. So we can support non-string keys.